include/materials/material_utility.tcc

```diff
@@ -0,0 +1,221 @@
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute q
+  const double q = mpm::material::q(stress);
+
+  // Compute deviatoric stress
+  const auto deviatoric_stress = mpm::material::deviatoric_stress(stress);
+
+  // Compute dq / dsigma
+  Eigen::Matrix<double, 6, 1> dq_dsigma;
+  dq_dsigma.setZero();
+  if (std::abs(q) > std::numeric_limits<double>::epsilon()) {
+    dq_dsigma(0) = 3. / (2. * q) * deviatoric_stress(0);
+    dq_dsigma(1) = 3. / (2. * q) * deviatoric_stress(1);
+    dq_dsigma(2) = 3. / (2. * q) * deviatoric_stress(2);
+    dq_dsigma(3) = 3. / q * deviatoric_stress(3);
+    dq_dsigma(4) = 3. / q * deviatoric_stress(4);
+    dq_dsigma(5) = 3. / q * deviatoric_stress(5);
+  }
+
+  return dq_dsigma;
+}
+
+//! Compute derivative of J2 in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1> mpm::material::dj2_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress and put them as jd2 / dsigma
+  auto dj2_dsigma = mpm::material::deviatoric_stress(stress);
+  dj2_dsigma(3) *= 2.0;
+  dj2_dsigma(4) *= 2.0;
+  dj2_dsigma(5) *= 2.0;
+
+  return dj2_dsigma;
+}
+
+//! Compute derivative of J3 in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1> mpm::material::dj3_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2
+  const double j2 = mpm::material::j2(stress);
+
+  // Compute deviatoric stress
+  const auto deviatoric_stress = mpm::material::deviatoric_stress(stress);
+
+  // Compute dj3 / dsigma
+  Eigen::Matrix<double, 3, 1> dev1;
+  dev1(0) = deviatoric_stress(0);
+  dev1(1) = deviatoric_stress(3);
+  dev1(2) = deviatoric_stress(5);
+  Eigen::Matrix<double, 3, 1> dev2;
+  dev2(0) = deviatoric_stress(3);
+  dev2(1) = deviatoric_stress(1);
+  dev2(2) = deviatoric_stress(4);
+  Eigen::Matrix<double, 3, 1> dev3;
+  dev3(0) = deviatoric_stress(5);
+  dev3(1) = deviatoric_stress(4);
+  dev3(2) = deviatoric_stress(2);
+
+  Eigen::Matrix<double, 6, 1> dj3_dsigma;
+  dj3_dsigma.setZero();
+  dj3_dsigma(0) = dev1.dot(dev1) - (2. / 3.) * j2;
+  dj3_dsigma(1) = dev2.dot(dev2) - (2. / 3.) * j2;
+  dj3_dsigma(2) = dev3.dot(dev3) - (2. / 3.) * j2;
+  dj3_dsigma(3) = 2.0 * dev1.dot(dev2);
+  dj3_dsigma(4) = 2.0 * dev2.dot(dev3);
+  dj3_dsigma(5) = 2.0 * dev1.dot(dev3);
+
+  return dj3_dsigma;
+}
+
+//! Compute derivative of Lode angle theta in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1> mpm::material::dtheta_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2
+  const double j2 = mpm::material::j2(stress);
+
+  // Compute J3
+  const double j3 = mpm::material::j3(stress);
+
+  // Compute dj2_dsigma
+  const auto dj2_dsigma = mpm::material::dj2_dsigma(stress);
+
+  // Compute dj3_dsigma
+  const auto dj3_dsigma = mpm::material::dj3_dsigma(stress);
+
+  // Declare R as zero to avoid division by zero J2
+  // R is defined as R = cos(3 theta)
+  double r = 0.0;
```

Comment:
Why are we declaring r here? Couldn't we declare it line 204?

Reply:
We can't for two reasons: (1) r is zero when j2 is zero (division by zero) so the same as the next three variables, we need to declare here and then update, (2) cannot declare within if statements

Reply:
But you are not using `r` anywhere else outside the loop starting in line 202. Or am I missing something?

Reply:
Sorry, we can!